### DISCUSSION TITLE
Method overloading problem

### DISCUSSION THREAD

**Discussion by rezayi** [caoccao/Javet#182] [Q&A] [answered]
Hi.
I have a problem with method overloading in javet.
in js we can define a method and call it with some of inputs. So other input parameters will set as undefined.
for example:
```
function a(x,y,z){
    console.log(x);
    console.log(y);
    console.log(z);
}
```
we can call function a as below:
```
a(1);
a(1,2)
a(1,2,3)
```
but the problem is when function a is defined in java. if parameters count of function call not equals with java function input parameters count it will throw an error. 
The solution can be method Overloading in java, but just first method of methods with same name will perform and other will be ignored. 
The final solution I found is doing such as javascript,I mean if function called with a part of input parameters we call other parameters as null .

In class V8FuntionCallback at method receiveCallback this change can resolve this problem
```
 if (args != null) {
     final int length = args.getLength();
     for (int i = 0; i < length; ++i) {
           values.add(args.get(i));
     }
}
```
to
```
 if (args != null) {
     final int length = args.getLength();
     for (int i = 0; i < length; ++i) {
           values.add(args.get(i));
     }
     while (values.size() < method.getParameterCount()) {
        values.add(null);
     }
}
```

**Comment by caoccao** [comment, marked as answer]
Please check this [doc](https://www.caoccao.com/Javet/reference/v8_values/v8_function.html#how-to-avoid-argument-type-or-count-mismatches) out.

**Reply by rezayi** [reply]
Thank you for the reply.
I agree that varargs could be a solution. But you must use varArgs for any function that has optional input parameters. But this solution reduces the readability of the code and makes it harder to use the input parameters because you have to extract the input parameters from the array.
But declaring the input parameters is clearer and more readable, and with my solution you can clearly define all the parameters and be sure that no errors will occur.

**Reply by caoccao** [reply]
The idea is fine. Here are the problems.

1. The signatures could be `f(..., int, long, boolean, ...)` so that filling `null` would cause type cast exception. The one I recommended is free from that.
2. Parameter count is not reliable when varargs is in the signature.
3. It still cannot resolve argument count exceeding expected count. The one I recommended is able to take redundant arguments without any problems.

**Reply by rezayi** [reply]
Regarding case 1, I should say that when you set a parameter as optional, you should not use primitive types. In fact, this situation is a bug. The problem can be solved by specifying Object instead of primitive.
Regarding item 2, when vararg really needed we can use your solution.
and Regarding item 3, I should say that our function is specified in Java and Java is a static language. So you can define all required parameters in Java and make sure that no more parameters will receive.
And finally, for exceeding parameters, you can do almost like JavaScript. when exceeding parameters passed to a function javascript will accept them but implemented code ignores them.when we implement a method in java we don't expect exceeding parameters. so we can ignore exceeding parameters when calling a Java method.
Finally, it seems that the best solution is to have both ways.

**Reply by caoccao** [reply]
That makes sense. I'll think it over.

**Reply by caoccao** [reply]
Maybe Strict Mode vs. Loose Mode, so that both of the Javet users can be satisfied.

**Reply by caoccao** [reply]
I decided to support it without a switch. Please check this [commit](https://github.com/caoccao/Javet/commit/5f01f5ce536de17b344a0f63d6a4a19ed9c44dfe) out. It will be included in the next release.

**Reply by rezayi** [reply]
Thank you for your attention. :)
